File: Assignment 2/task2.scala
import org.apache.spark.{SparkConf, SparkContext}

import java.io.{File, PrintWriter}
import scala.collection.mutable
import scala.util.control.Breaks.{break, breakable}
import scala.collection.mutable.{Set => MutableSet}
import scala.math.Ordering.Implicits._
import scala.collection.mutable.{ListBuffer, Set => MutableSet}

object task2 extends Serializable{

  def find_candidate_itemsets(freq_item_sets: ListBuffer[Set[String]]) : ListBuffer[Set[String]] = {
    val res: MutableSet[Set[String]] = MutableSet()
    val candidate_item_set_length = freq_item_sets.head.size + 1

    val n = freq_item_sets.length
    for (i <- 0 to n - 2){
      for (j <- i + 1 until n){
        val new_item_set = freq_item_sets(i).union(freq_item_sets(j))
        if (new_item_set.size == candidate_item_set_length){
          res += new_item_set
        }
      }
    }
    res.to[mutable.ListBuffer]
  }

  def find_frequent_itemset(candidate_item_sets: mutable.ListBuffer[Set[String]], baskets: Array[Set[String]], partial_support: Float): mutable.ListBuffer[Set[String]] = {
    val res = mutable.ListBuffer[Set[String]]()
    for (item_set <- candidate_item_sets){
      var counter = 0
      for (b <- baskets){
        if (item_set.subsetOf(b)){
          counter = counter + 1
        }
      }

      if (counter >= partial_support){
        res += item_set
      }
    }
    res
  }

  def find_frequent_itemsets(baskets: Array[Set[String]], support: Int, totalBaskets: Long): Iterator[Set[String]] = {
    var res            = mutable.ListBuffer[Set[String]]()
    val count_map      = mutable.HashMap[String, Int]()
    var freq_item_sets = mutable.ListBuffer[Set[String]]()

    for (b <- baskets) {
      for (i <- b) {
        count_map(i) = count_map.getOrElse(i, 0) + 1
      }
    }

    val partial_support = (support.toFloat * baskets.length) / totalBaskets
    for ((k, v) <- count_map){
      if (v >= partial_support){
        freq_item_sets += Set(k)
        res += Set(k)
      }
    }

    breakable {
      while(freq_item_sets.nonEmpty){
        val candidate_item_sets = find_candidate_itemsets(freq_item_sets)
        if(candidate_item_sets.nonEmpty){
          freq_item_sets = find_frequent_itemset(candidate_item_sets, baskets, partial_support)
          if(freq_item_sets.nonEmpty){
            // concatenate
            res = res ++ freq_item_sets
          }
        }else{
          // while loop break condition
          break
        }
      }
    }

    res.toIterator
  }

  def get_item_count_per_candidate(baskets: Array[Set[String]], candidate_sets: Array[Set[String]]): Iterator[(Set[String], Long)] = {
    val res = mutable.ListBuffer[(Set[String], Long)]()
    for(item_sets <- candidate_sets){
      var cnt = 0
      for(b <- baskets){
        if(item_sets.subsetOf(b)){
          cnt = cnt + 1
        }
      }
      res += Tuple2(item_sets, cnt)
    }

    res.toIterator
  }


  def format_output(item_sets: Array[Set[String]], writer: PrintWriter) :Unit = {
    val sorted_grouped_item_sets = item_sets.groupBy(l => l.size).toList.sortBy(_._1)
    for ((_, item_sets) <- sorted_grouped_item_sets){
      val sorted_item_sets = item_sets.map(item_set => item_set.toList.sorted).sorted.map(item_set => "(" + item_set.map(i => "'" + i + "'").mkString(", ") + ")").mkString(",")
      writer.write(sorted_item_sets + "\n\n")
    }
  }

  def write_output(candidate_sets: Array[Set[String]], frequent_sets: Array[Set[String]], output_file: String): Unit = {

    val writer = new PrintWriter(new File(output_file))
    writer.write("Candidates:\n")
    format_output(candidate_sets, writer)
    writer.write("Frequent Itemsets:\n")
    format_output(frequent_sets, writer)
    writer.close()
  }

  def task2_son(filter_threshold: Int, support: Int, input_file : String, output_file: String, sc : SparkContext) : Unit = {

    var input_data_rdd = sc.textFile(input_file).map(l => l.split(','))
    val header         = input_data_rdd.first()
    input_data_rdd     = input_data_rdd.filter(l => l != header)
    // Basket Creation
    // Combinations of frequent businesses
    // initialize baskets with entire rdd and then create baskets
    var baskets =  input_data_rdd.map(l => (l(0), l(1)))
    var buckets = baskets.groupByKey().map(l => l._2.toSet)
    buckets     = buckets.filter(l => l.size > filter_threshold)
    val totalBaskets = buckets.count()
    val candidate_sets = buckets.mapPartitions(l => find_frequent_itemsets(l.toArray, support, totalBaskets)).distinct().collect()
    val frequent_sets = buckets.mapPartitions(l => get_item_count_per_candidate(l.toArray, candidate_sets)).reduceByKey((a,b)=> a+b).filter(l=>l._2 >= support).map(l=>l._1).collect()

    write_output(candidate_sets, frequent_sets, output_file)

  }


  def format_csv_dataset(sc: SparkContext, input_file: String, formatted_csv_file: String): Unit = {
    var input_data_rdd = sc.textFile(input_file)
    val header_data = input_data_rdd.first()
    input_data_rdd = input_data_rdd.filter(l => l != header_data)
    val formatted_data = input_data_rdd.map(data => (data.split(",")(0).stripPrefix("\"").stripSuffix("\"") + "-" + data.split(",")(1).stripPrefix("\"").stripSuffix("\""), data.split(",")(5).stripPrefix("\"").stripSuffix("\"").toLong)).collect()

    val writer = new PrintWriter(new File(formatted_csv_file))
    writer.write("DATE-CUSTOMER_ID,PRODUCT_ID\n")
    for(row <- formatted_data) {
      writer.write(row._1 + "," + row._2 + "\r\n")
    }
    writer.close()
  }

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("DM_assign2_task2").setMaster("local[*]")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    val filter_threshold = args(0).toInt
    val support = args(1).toInt
    val input_file = args(2)
    val output_file = args(3)
    val formatted_csv_file = "customer_product_file.csv"
    format_csv_dataset(sc, input_file, formatted_csv_file)
    val startTime      = System.nanoTime()
    task2_son(filter_threshold, support, formatted_csv_file, output_file ,sc)
    println("Duration: " + (System.nanoTime - startTime) / 1e9d)
  }
}

File: Assignment 2/task2.py
# POINTERS FOR ASSIGNMENT TASK2
# Exe time starts from loading preprocessed data.


import csv, sys, time
from pyspark import SparkContext, SparkConf
from collections import defaultdict

_FORMATTED_CSV_FILE = 'customer_product_file.csv'

def create_date_customer_id(row):
    row_list  = row.split(',')
    date_str  = row_list[0]
    date_str  = date_str.replace('"', '')
    year      = date_str[-2:]
    date_list = date_str.split('/')
    formatted_date = '/'.join([date_list[0], date_list[1], year])
    return formatted_date + '-' + row_list[1].replace('"', '').lstrip('0')

def create_product_id(row):
    
    product_id = row.split(',')[5]
    return int(product_id.replace('"','').lstrip('0'))

    
def format_csv_dataset(sc, ip_file, outfile):
    #ip_file, outfile = 'ta_feng_all_months_merged.csv','task2.csv'
    input_data_rdd = sc.textFile(ip_file)
    header         = input_data_rdd.first()
    input_data_rdd = input_data_rdd.filter(lambda x: x!=header)
    formatted_csv  = input_data_rdd.map(lambda row: (create_date_customer_id(row), create_product_id(row)))
    with open(outfile, 'w', newline='') as f:
        writer = csv.writer(f, delimiter=',',quoting=csv.QUOTE_NONE)
        writer.writerow(["DATE-CUSTOMER_ID", "PRODUCT_ID"])
        writer.writerows(formatted_csv.collect())

def get_item_count_per_candidate(baskets,candidate_sets):
    counts_dict = defaultdict(int)
    baskets      = list(baskets)
    for i in candidate_sets:
        for b in baskets:
            if i.issubset(b):
                counts_dict[i] += 1
    return counts_dict.items()

def find_candidate_itemset(frequent_items,combination):
    frequent_items_length = len(frequent_items)
    candidate_items_set = set()
    for i in range(frequent_items_length-1):
        for j in range(i+1,frequent_items_length):
            candidate_set = frequent_items[i].union(frequent_items[j])
            if len(candidate_set) == combination:
                candidate_items_set.add(candidate_set)
    return candidate_items_set

def populate_frequent_items_list(counts_dict, frequent_items, all_frequent_items_list, ps):
    for k,v in counts_dict.items(): 
        if v < ps:
            continue
        frequent_items.append(k)
        all_frequent_items_list.append((k,1))
    return frequent_items, all_frequent_items_list

def find_frequent_itemset(candidate_items_set, basket_list,ps, all_frequent_items_list):
    counts_dict = defaultdict(int)
    for i in candidate_items_set:
        for j in basket_list:
            if i.issubset(j):
                counts_dict[i] += 1
    
    frequent_items = []
    frequent_items, all_frequent_items_list = populate_frequent_items_list(counts_dict, frequent_items, all_frequent_items_list, ps)
    return frequent_items,all_frequent_items_list

def find_frequent_itemsets(baskets, support, items_total):
    baskets_list            = [] 
    all_frequent_items_list = []
    frequent_items          = []
    counts_dict             = defaultdict(int)
    combinations            = 2

    for b in baskets:
        baskets_list.append(b)
        for i in b:
            counts_dict[ frozenset( (i,) ) ] += 1

    ps = support * (len(baskets_list) / items_total)
    
    frequent_items, all_frequent_items_list = populate_frequent_items_list(counts_dict, frequent_items, all_frequent_items_list, ps)

    while True:
        candidate_items_list                         = find_candidate_itemset(frequent_items,combinations)
        frequent_items, all_frequent_items_list      = find_frequent_itemset(candidate_items_list, baskets_list,ps, all_frequent_items_list)

        if len(candidate_items_list) == 0 or len(frequent_items) == 0:
            break
        combinations += 1
    return all_frequent_items_list

def format_output(items_set,f):
    itemset_length_dict = defaultdict(lambda : [])
    for i in items_set:
        itemset_length_dict[len(i)].append(sorted(tuple(i)))
    for i in itemset_length_dict.values():
        i.sort()
        f.write(','.join(list(map(lambda val: str(tuple(val)).replace(',)',')'),i))))
        f.write('\n\n')

def write_output(candidate_sets,frequent_sets, output_file):
    with open(output_file,'w+') as f:
        f.write('Candidates:\n')
        format_output(candidate_sets,f)
        f.write('Frequent Itemsets:\n')
        format_output(frequent_sets,f)

def task2(input_data_rdd, support, filter_threshold, op_file):
    header_data    = input_data_rdd.first()
    input_data_rdd = input_data_rdd.filter(lambda row: row != header_data)
    baskets        = input_data_rdd.filter(lambda x: x != header_data)\
                                   .map(lambda x : (x[0],frozenset( [x[1]] ) ))\
                                   .reduceByKey(lambda x,y : x | y)\
                                   .filter(lambda x:len(x[1]) > filter_threshold)\
                                   .map(lambda x : x[1])
    print(baskets.first())

    item_total     = baskets.count()
    candidate_sets = baskets.mapPartitions(lambda row: find_frequent_itemsets(row,support,item_total))\
                            .groupByKey()\
                            .map(lambda x: x[0])\
                            .collect()

    frequent_sets = baskets.mapPartitions(lambda row : get_item_count_per_candidate(row,candidate_sets))\
                           .reduceByKey(lambda x,y: x + y)\
                           .filter(lambda row: row[1] >= support)\
                           .map(lambda row :list(row[0]))\
                           .collect()
    write_output(candidate_sets, frequent_sets, op_file)

def main():
    conf = SparkConf().setAppName('DM_assign2_task2').setMaster('local[*]')
    sc = SparkContext.getOrCreate(conf=conf)
    sc.setLogLevel("ERROR")
    sc.setSystemProperty('spark.driver.memory', '4g')
    sc.setSystemProperty('spark.executor.memory', '4g')
    args = sys.argv
    filter_threshold, support, ip_file, op_file = int(args[1]), int(args[2]), str(args[3]), str(args[4])
    format_csv_dataset(sc, ip_file, _FORMATTED_CSV_FILE)
    st_time = time.time()
    #Read this generated csv and then apply SON on it.
    formatted_rdd = sc.textFile(_FORMATTED_CSV_FILE)\
                       .map(lambda row : row.split(','))
    print(formatted_rdd.first())
    task2(formatted_rdd, support, filter_threshold, op_file)

    print('Duration: {0}'.format(time.time() - st_time))

main()